File: python/src/mpeg_o/enums.py
"""Enumerations mirroring the Objective-C ``MPGOEnums.h`` integer values.

All enums are ``IntEnum`` subclasses because every enum in the format
is persisted on disk as an integer attribute (for example
``@acquisition_mode = 0``). Keeping the Python-side values identical
to the Objective-C ``NS_ENUM`` values makes the HDF5 layer a direct
pass-through and removes any translation table to go stale.

Cross-language equivalents
--------------------------
Objective-C: ``MPGOEnums.h`` ·
Java: ``com.dtwthalion.mpgo.Enums``
"""
from __future__ import annotations

from enum import IntEnum


class SamplingMode(IntEnum):
    """Axis sampling regularity.

    Cross-language: ObjC ``MPGOSamplingMode`` · Java
    ``Enums.SamplingMode``.
    """

    UNIFORM = 0
    NON_UNIFORM = 1


class Precision(IntEnum):
    """Numeric precision of a signal buffer.

    Cross-language: ObjC ``MPGOPrecision`` · Java
    ``Enums.Precision``.
    """

    FLOAT32 = 0
    FLOAT64 = 1
    INT32 = 2
    INT64 = 3
    UINT32 = 4
    COMPLEX128 = 5

    def numpy_dtype(self) -> str:
        """Return the little-endian NumPy dtype string for this precision."""
        return {
            Precision.FLOAT32: "<f4",
            Precision.FLOAT64: "<f8",
            Precision.INT32: "<i4",
            Precision.INT64: "<i8",
            Precision.UINT32: "<u4",
            Precision.COMPLEX128: "<c16",
        }[self]


class Compression(IntEnum):
    """Compression algorithm applied to a signal buffer.

    Cross-language: ObjC ``MPGOCompression`` · Java
    ``Enums.Compression``.
    """

    NONE = 0
    ZLIB = 1
    LZ4 = 2
    NUMPRESS_DELTA = 3


class ByteOrder(IntEnum):
    """Byte order of a signal buffer on disk.

    Cross-language: ObjC ``MPGOByteOrder`` · Java
    ``Enums.ByteOrder``.
    """

    LITTLE_ENDIAN = 0
    BIG_ENDIAN = 1


class Polarity(IntEnum):
    """Ion polarity for mass spectrometry.

    Cross-language: ObjC ``MPGOPolarity`` · Java
    ``Enums.Polarity``.
    """

    UNKNOWN = 0
    POSITIVE = 1
    NEGATIVE = -1


class ChromatogramType(IntEnum):
    """Chromatogram kind.

    Cross-language: ObjC ``MPGOChromatogramType`` · Java
    ``Enums.ChromatogramType``.
    """

    TIC = 0
    XIC = 1
    SRM = 2


class AcquisitionMode(IntEnum):
    """High-level acquisition scheme for a run.

    Cross-language: ObjC ``MPGOAcquisitionMode`` · Java
    ``Enums.AcquisitionMode``.
    """

    MS1_DDA = 0
    MS2_DDA = 1
    DIA = 2
    SRM = 3
    NMR_1D = 4
    NMR_2D = 5
    IMAGING = 6


class EncryptionLevel(IntEnum):
    """Multi-level protection granularity (MPEG-G style).

    Cross-language: ObjC ``MPGOEncryptionLevel`` · Java
    ``Enums.EncryptionLevel``.
    """

    NONE = 0
    DATASET_GROUP = 1
    DATASET = 2
    DESCRIPTOR_STREAM = 3
    ACCESS_UNIT = 4

File: java/src/main/java/com/dtwthalion/mpgo/Enums.java
/*
 * MPEG-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package com.dtwthalion.mpgo;

/**
 * Enumerations mirroring the ObjC {@code MPGOEnums.h} integer values.
 *
 * <p>Declaration order in each {@code enum} matches the ObjC
 * {@code NS_ENUM} integer values so that {@link Enum#ordinal()} is a
 * direct pass-through of the on-disk attribute. No translation table
 * can go stale.</p>
 *
 * <p><b>API status:</b> Stable.</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code MPGOEnums.h}, Python {@code mpeg_o.enums}.</p>
 *
 * @since 0.6
 */
public final class Enums {

    private Enums() {}

    /** Numeric precision of a signal buffer.
     *
     * <p><b>Appendix B Gap 7:</b> this enum used to hold a
     * {@code HDF5Constants.H5T_NATIVE_*} id per constant, which made
     * the HDF5 JNI wrapper a load-time dependency of every consumer,
     * including non-HDF5 providers such as SQLite. The HDF5 type ids
     * have moved to {@code Hdf5Group.hdf5TypeFor(Precision)} so this
     * enum no longer pulls in the native library.</p>
     */
    public enum Precision {
        /** 32-bit IEEE 754 single-precision float (4 bytes). */
        FLOAT32(4),
        /** 64-bit IEEE 754 double-precision float (8 bytes). */
        FLOAT64(8),
        /** Signed 32-bit integer (4 bytes). */
        INT32(4),
        /** Signed 64-bit integer (8 bytes). */
        INT64(8),
        /** Unsigned 32-bit integer stored as signed {@code int} (4 bytes). */
        UINT32(4),
        /** 128-bit complex: two {@code double} values per element (16 bytes). */
        COMPLEX128(16);

        private final int elementSize;

        Precision(int elementSize) {
            this.elementSize = elementSize;
        }

        /** @return size in bytes of a single element at this precision. */
        public int elementSize() { return elementSize; }
    }

    /** Compression algorithm applied to a signal buffer. */
    public enum Compression {
        /** No compression; raw binary layout. */
        NONE,
        /** Deflate (zlib/gzip-compatible) compression. */
        ZLIB,
        /** LZ4 block compression. */
        LZ4,
        /** NumPRESS delta-integer compression for ordered m/z or retention-time arrays. */
        NUMPRESS_DELTA
    }

    /** Ion polarity for mass spectrometry. */
    public enum Polarity {
        /** Polarity not specified or not applicable. */
        UNKNOWN(0),
        /** Positive-ion mode. */
        POSITIVE(1),
        /** Negative-ion mode. */
        NEGATIVE(-1);

        private final int value;
        Polarity(int value) { this.value = value; }

        /** @return the integer on-disk representation of this polarity. */
        public int intValue() { return value; }

        /**
         * Resolve an integer on-disk value to a {@link Polarity} constant.
         *
         * @param v the integer value read from the file
         * @return the matching constant, or {@link #UNKNOWN} if unrecognised
         */
        public static Polarity fromInt(int v) {
            return switch (v) {
                case 1 -> POSITIVE;
                case -1 -> NEGATIVE;
                default -> UNKNOWN;
            };
        }
    }

    /** Axis sampling regularity. */
    public enum SamplingMode {
        /** Evenly-spaced samples (fixed step size). */
        UNIFORM,
        /** Arbitrarily-spaced samples (coordinate array required). */
        NON_UNIFORM
    }

    /** High-level acquisition scheme for a run. */
    public enum AcquisitionMode {
        /** Data-dependent acquisition, MS1 survey scan. */
        MS1_DDA,
        /** Data-dependent acquisition, MS2 fragmentation scan. */
        MS2_DDA,
        /** Data-independent acquisition. */
        DIA,
        /** Selected reaction monitoring. */
        SRM,
        /** One-dimensional NMR experiment. */
        NMR_1D,
        /** Two-dimensional NMR experiment. */
        NMR_2D,
        /** Mass spectrometry imaging. */
        IMAGING
    }

    /** Chromatogram kind. */
    public enum ChromatogramType {
        /** Total ion chromatogram. */
        TIC,
        /** Extracted ion chromatogram. */
        XIC,
        /** Selected reaction monitoring chromatogram. */
        SRM
    }

    /** Byte order of a signal buffer on disk. */
    public enum ByteOrder {
        /** Least-significant byte first (x86 native). */
        LITTLE_ENDIAN,
        /** Most-significant byte first (network order). */
        BIG_ENDIAN
    }

    /** Multi-level protection granularity (MPEG-G style). */
    public enum EncryptionLevel {
        /** No encryption applied. */
        NONE,
        /** Protection spans a complete dataset group. */
        DATASET_GROUP,
        /** Protection spans a single dataset. */
        DATASET,
        /** Protection spans a descriptor stream within a dataset. */
        DESCRIPTOR_STREAM,
        /** Protection spans a single access unit (finest granularity). */
        ACCESS_UNIT
    }
}
